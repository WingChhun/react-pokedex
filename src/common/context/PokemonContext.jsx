import React, { createContext, useReducer, useEffect } from 'react';
import { SELECT_POKEMON, READ_POKEMON } from '../constants';
import { fetchPokemon } from '../api';
import { pokemonReducer } from './reducers';
import LocalStorageMgr from '../LocalStorageMgr';

const LIMIT = 151;
const PokemonContext = createContext();

const PokemonProvider = ({ children }) => {
  const [state, dispatch] = useReducer(pokemonReducer, {
    filterStr: '',
    showSaved: false,
    pokemon: {},
    savedPokemon: [],
    selected: LocalStorageMgr.getReducer(SELECT_POKEMON) || {}
  });

  useEffect(() => {
    const cachedPokemon = LocalStorageMgr.getReducer(READ_POKEMON) || {};

    async function getPokemon() {
      try {
        const { results } = await fetchPokemon(LIMIT);
        const mappedByIndex = results.map((pokemon, spriteId) => ({
          ...pokemon,
          spriteId: spriteId + 1
        }));

        dispatch({
          type:READ_POKEMON,
          payload: mappedByIndex
        });

        LocalStorageMgr.setReducer(READ_POKEMON, mappedByIndex);
      } catch (err) {
        console.warn('Error in PokemonContext fetching request', err);
      }
    }

    //Call fetchPokemon if none are cached
    Object.keys(cachedPokemon).length === 0
      ? getPokemon()
      : dispatch({ type: 'READ_POKEMON', payload: cachedPokemon });
  }, []);

  return (
    <PokemonContext.Provider value={[state, dispatch]}>
      {children}
    </PokemonContext.Provider>
  );
};

export { PokemonProvider, PokemonContext };
