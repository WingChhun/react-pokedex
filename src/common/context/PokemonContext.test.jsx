import React from 'react';
import {PokemonProvider, PokemonContext} from "./PokemonContext";
import { configure,shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe(`<PokemonContext/>`, () => {
const props = {
    children: jest.fn()
};

  it(`renders properly`, () => {
    const wrapper = mount(<PokemonContext.Consumer {...props}/>);
    expect(wrapper).toMatchSnapshot();
  });


  it(`tests useEffect`,()=>{
    const wrapper = mount(<PokemonContext.Consumer {...props}/>)
    console.log(wrapper.debug());
console.log("PokemonContext", PokemonContext.Consumer);


  });
});