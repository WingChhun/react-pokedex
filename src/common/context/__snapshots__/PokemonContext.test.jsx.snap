// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonContext/> renders properly 1`] = `ReactWrapper {}`;
